# Neste notebook vamos fazer o tratamento e limpeza dos dados.

## Analise de Dados de Airbnb no Rio de Janeiro 

Fonte: Inside Airbnb - Rio de Janeiro - https://insideairbnb.com/rio-de-janeiro/

## 1. Carregamento de Dados

In [4]:
# importando bibliotecas 
import polars as pl
import matplotlib.pyplot as plt
import pandas as pd
import seaborn as sns
import numpy as np

# remover avisos desnecessarios
import warnings
warnings.filterwarnings('ignore')

In [5]:
# carrega os dados
df = pd.read_csv('listings_rio_de_janeiro_cleaned.csv')

## 2. Limpeza e tratamento dos dados.

In [6]:
# visualizando as 5 primeiras linhas
df.head()

,neighbourhood_cleansed,latitude,longitude,property_type,room_type,accommodates,bathrooms,bathrooms_text,bedrooms,beds,...,review_scores_accuracy,review_scores_cleanliness,review_scores_checkin,review_scores_communication,review_scores_location,review_scores_value,instant_bookable,reviews_per_month,num_bathrooms,price_per_person
0,Leblon,-22.982818,-43.222457,Entire condo,Entire home/apt,4,1.5,1.5 baths,2.0,3.0,...,4.97,4.99,4.97,4.94,4.99,4.85,f,2.72,1.5,145.000000
1,Copacabana,-22.984090,-43.191770,Entire rental unit,Entire home/apt,2,2.0,2 baths,1.0,1.0,...,4.93,4.90,4.98,4.99,4.96,4.83,f,0.74,2.0,950.000000
2,Pavuna,-22.814911,-43.379011,Entire rental unit,Entire home/apt,4,1.0,1 bath,1.0,1.0,...,4.93,4.90,4.98,4.99,4.96,4.83,t,0.74,1.0,175.000000
3,Barra da Tijuca,-23.010000,-43.344820,Entire rental unit,Entire home/apt,4,1.0,1 bath,1.0,1.0,...,5.00,4.82,5.00,5.00,5.00,5.00,f,0.60,1.0,125.000000
4,Copacabana,-22.970696,-43.186048,Entire rental unit,Entire home/apt,6,2.0,2 baths,2.0,6.0,...,4.93,4.88,4.94,4.99,4.94,4.85,t,2.81,2.0,76.833333


In [7]:
# visualizando as 5 ultimas linhas
df.tail()

,neighbourhood_cleansed,latitude,longitude,property_type,room_type,accommodates,bathrooms,bathrooms_text,bedrooms,beds,...,review_scores_accuracy,review_scores_cleanliness,review_scores_checkin,review_scores_communication,review_scores_location,review_scores_value,instant_bookable,reviews_per_month,num_bathrooms,price_per_person
38665,Ipanema,-22.987120,-43.192690,Entire rental unit,Entire home/apt,2,1.0,1 bath,1.0,1.0,...,4.93,4.9,4.98,4.99,4.96,4.83,t,0.74,1.0,282.5
38666,Botafogo,-22.952837,-43.194664,Private room in home,Private room,2,1.0,1 shared bath,1.0,1.0,...,4.93,4.9,4.98,4.99,4.96,4.83,f,0.74,1.0,47.5
38667,Copacabana,-22.967640,-43.183030,Entire rental unit,Entire home/apt,2,1.0,1 bath,1.0,1.0,...,4.93,4.9,4.98,4.99,4.96,4.83,f,0.74,1.0,102.0
38668,Botafogo,-22.951844,-43.194001,Private room in home,Private room,1,0.5,Private half-bath,1.0,2.0,...,4.93,4.9,4.98,4.99,4.96,4.83,t,0.74,NaN,112.0
38669,Laranjeiras,-22.939740,-43.190110,Private room in home,Private room,2,1.5,1.5 baths,1.0,1.0,...,4.93,4.9,4.98,4.99,4.96,4.83,f,0.74,1.5,73.0


In [8]:
# visualizando os tipos de dados
df.dtypes

neighbourhood_cleansed             str
latitude                       float64
longitude                      float64
property_type                      str
room_type                          str
accommodates                     int64
bathrooms                      float64
bathrooms_text                     str
bedrooms                       float64
beds                           float64
amenities                          str
price                          float64
minimum_nights                   int64
maximum_nights                   int64
availability_30                  int64
availability_60                  int64
availability_90                  int64
availability_365                 int64
number_of_reviews                int64
review_scores_rating           float64
review_scores_accuracy         float64
review_scores_cleanliness      float64
review_scores_checkin          float64
review_scores_communication    float64
review_scores_location         float64
review_scores_value      

In [9]:
# visualizando os nulos
df.isnull().sum().sort_values(ascending=False)

num_bathrooms                  266
bathrooms                       39
longitude                        0
property_type                    0
room_type                        0
neighbourhood_cleansed           0
accommodates                     0
bathrooms_text                   0
bedrooms                         0
beds                             0
amenities                        0
price                            0
minimum_nights                   0
latitude                         0
maximum_nights                   0
availability_30                  0
availability_90                  0
availability_60                  0
number_of_reviews                0
review_scores_rating             0
review_scores_accuracy           0
availability_365                 0
review_scores_cleanliness        0
review_scores_checkin            0
review_scores_location           0
review_scores_communication      0
review_scores_value              0
instant_bookable                 0
reviews_per_month   

In [10]:
# preenchendo valores nulos
df = df.fillna({'num_bathrooms': 0, 'bathrooms': 0})

In [11]:
# verificando linhas duplicadas
df.drop(columns='neighbourhood_cleansed').duplicated().sum()

np.int64(1)

In [12]:
# removendo linhas duplicadas 
df.drop_duplicates()

,neighbourhood_cleansed,latitude,longitude,property_type,room_type,accommodates,bathrooms,bathrooms_text,bedrooms,beds,...,review_scores_accuracy,review_scores_cleanliness,review_scores_checkin,review_scores_communication,review_scores_location,review_scores_value,instant_bookable,reviews_per_month,num_bathrooms,price_per_person
0,Leblon,-22.982818,-43.222457,Entire condo,Entire home/apt,4,1.5,1.5 baths,2.0,3.0,...,4.97,4.99,4.97,4.94,4.99,4.85,f,2.72,1.5,145.000000
1,Copacabana,-22.984090,-43.191770,Entire rental unit,Entire home/apt,2,2.0,2 baths,1.0,1.0,...,4.93,4.90,4.98,4.99,4.96,4.83,f,0.74,2.0,950.000000
2,Pavuna,-22.814911,-43.379011,Entire rental unit,Entire home/apt,4,1.0,1 bath,1.0,1.0,...,4.93,4.90,4.98,4.99,4.96,4.83,t,0.74,1.0,175.000000
3,Barra da Tijuca,-23.010000,-43.344820,Entire rental unit,Entire home/apt,4,1.0,1 bath,1.0,1.0,...,5.00,4.82,5.00,5.00,5.00,5.00,f,0.60,1.0,125.000000
4,Copacabana,-22.970696,-43.186048,Entire rental unit,Entire home/apt,6,2.0,2 baths,2.0,6.0,...,4.93,4.88,4.94,4.99,4.94,4.85,t,2.81,2.0,76.833333
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
38665,Ipanema,-22.987120,-43.192690,Entire rental unit,Entire home/apt,2,1.0,1 bath,1.0,1.0,...,4.93,4.90,4.98,4.99,4.96,4.83,t,0.74,1.0,282.500000
38666,Botafogo,-22.952837,-43.194664,Private room in home,Private room,2,1.0,1 shared bath,1.0,1.0,...,4.93,4.90,4.98,4.99,4.96,4.83,f,0.74,1.0,47.500000
38667,Copacabana,-22.967640,-43.183030,Entire rental unit,Entire home/apt,2,1.0,1 bath,1.0,1.0,...,4.93,4.90,4.98,4.99,4.96,4.83,f,0.74,1.0,102.000000
38668,Botafogo,-22.951844,-43.194001,Private room in home,Private room,1,0.5,Private half-bath,1.0,2.0,...,4.93,4.90,4.98,4.99,4.96,4.83,t,0.74,0.0,112.000000


In [13]:
# verificando localização exata
colunas_geo = ['latitude', 'longitude']

total_duplicados_geo = df.duplicated(subset=colunas_geo).sum()

print(f"Existem {total_duplicados_geo} anúncios que compartilham a mesma localização exata.")


Existem 4259 anúncios que compartilham a mesma localização exata.


In [14]:
# filtrando o DataFrame para mostrar apenas as linhas que têm coordenadas repetidas
anuncios_no_mesmo_lugar = df[df.duplicated(subset=colunas_geo, keep=False)]

# ordena por latitude para ver os pares juntos
display(anuncios_no_mesmo_lugar.sort_values(by='latitude').head(10))


,neighbourhood_cleansed,latitude,longitude,property_type,room_type,accommodates,bathrooms,bathrooms_text,bedrooms,beds,...,review_scores_accuracy,review_scores_cleanliness,review_scores_checkin,review_scores_communication,review_scores_location,review_scores_value,instant_bookable,reviews_per_month,num_bathrooms,price_per_person
35977,Barra de Guaratiba,-23.065260,-43.565476,Shared room in hotel,Shared room,16,2.0,2 shared baths,1.0,22.0,...,4.93,4.90,4.98,4.99,4.96,4.83,f,0.74,2.0,34.5000
443,Barra de Guaratiba,-23.065260,-43.565476,Entire home,Entire home/apt,2,1.0,1 bath,1.0,1.0,...,4.93,4.90,4.98,4.99,4.96,4.83,f,0.74,1.0,100.0000
35979,Barra de Guaratiba,-23.065260,-43.565476,Room in hotel,Private room,16,2.0,2 shared baths,4.0,43.0,...,4.93,4.90,4.98,4.99,4.96,4.83,f,0.74,2.0,14.6875
22348,Barra de Guaratiba,-23.064618,-43.566976,Entire guesthouse,Entire home/apt,2,1.0,1 bath,1.0,4.0,...,5.00,4.94,5.00,5.00,5.00,5.00,f,1.19,1.0,70.0000
30741,Barra de Guaratiba,-23.064618,-43.566976,Entire home,Entire home/apt,10,3.0,3 baths,4.0,5.0,...,4.78,5.00,4.67,4.89,4.78,4.89,f,1.30,3.0,53.5000
18977,Barra de Guaratiba,-23.055640,-43.558204,Private room in home,Private room,2,1.0,1 private bath,1.0,1.0,...,5.00,5.00,5.00,5.00,4.89,4.89,f,0.40,1.0,100.0000
18981,Barra de Guaratiba,-23.055640,-43.558204,Private room in home,Private room,2,1.0,1 private bath,1.0,1.0,...,5.00,5.00,4.91,5.00,5.00,5.00,f,0.52,1.0,100.0000
33203,Recreio dos Bandeirantes,-23.033871,-43.491881,Shipping container,Entire home/apt,2,1.0,1 bath,1.0,1.0,...,4.50,4.50,5.00,5.00,4.50,4.00,t,0.38,1.0,91.0000
37354,Recreio dos Bandeirantes,-23.033871,-43.491881,Entire rental unit,Entire home/apt,2,1.0,1 bath,1.0,2.0,...,4.93,4.90,4.98,4.99,4.96,4.83,t,0.74,1.0,154.0000
15195,Barra de Guaratiba,-23.031819,-43.563896,Entire chalet,Entire home/apt,4,1.0,1 bath,1.0,2.0,...,4.93,4.90,4.98,4.99,4.96,4.83,t,0.74,1.0,75.0000


## Decisão sobre os dados com localização iguais

Identificamos que 4259 anúncios compartilham as localizações entre si. Vamos considerar mais de um imóvel por localização. 

# 3. Renomeando colunas

In [15]:
# 1. Renomeando colunas principais

df = df.rename(columns={'neighbourhood_cleansed': 'BAIRRO', 
                    'property_type': 'TIPO_IMOVEL',
                    'room_type': 'TIPO_QUARTO',
                    'bathrooms': 'BANHEIROS',
                    'bedrooms': 'QUARTOS',
                    'beds': 'CAMAS',
                    'amenities': 'COMODIDADE',
                    'price': 'PRECO',
                    'minimum_nights': 'NOITES_MINIMAS',
                    'maximum_nights': 'NOITES_MAXIMAS',
                    'accommodates': 'LIMITE_PESSOAS',
                    'review_scores_accuracy': 'MEDIA_AVALIACAO',
                    'review_scores_communication': 'AVALIACAO_COMUNICACAO',
                    'review_scores_location': 'AVALIACAO_LOCALIZACAO',
                    'price_per_person': 'PRECO_POR_PESSOA'
    })

# 2. Renomeando dados dentro das colunas
condicoes = [
    df['TIPO_IMOVEL'].str.contains(r'shared room|shared', case=False, na=False),
    df['TIPO_IMOVEL'].str.contains(r'private room', case=False, na=False),
    df['TIPO_IMOVEL'].str.contains(r'condo|unit|apartment|loft|suit', case=False, na=False),
    df['TIPO_IMOVEL'].str.contains(r'house|villa|entire home', case=False, na=False),
    df['TIPO_IMOVEL'].str.contains(r'room in hotel|Room in hostel', case=False, na=False)
]

escolhas = [
    'Quarto Compartilhado',
    'Quarto Privado',
    'Apartamento',
    'Casa',
    'Hotel/Hostel'
]

df['TIPO_IMOVEL'] = np.select(condicoes, escolhas, default='Outros')

In [16]:
# Arredondando para 2 casas decimais
df['PRECO_POR_PESSOA'] = df['PRECO_POR_PESSOA'].round(2)

In [18]:
df.head(10)

,BAIRRO,latitude,longitude,TIPO_IMOVEL,TIPO_QUARTO,LIMITE_PESSOAS,BANHEIROS,bathrooms_text,QUARTOS,CAMAS,...,MEDIA_AVALIACAO,review_scores_cleanliness,review_scores_checkin,AVALIACAO_COMUNICACAO,AVALIACAO_LOCALIZACAO,review_scores_value,instant_bookable,reviews_per_month,num_bathrooms,PRECO_POR_PESSOA
0,Leblon,-22.982818,-43.222457,Apartamento,Entire home/apt,4,1.5,1.5 baths,2.0,3.0,...,4.97,4.99,4.97,4.94,4.99,4.85,f,2.72,1.5,145.00
1,Copacabana,-22.984090,-43.191770,Apartamento,Entire home/apt,2,2.0,2 baths,1.0,1.0,...,4.93,4.90,4.98,4.99,4.96,4.83,f,0.74,2.0,950.00
2,Pavuna,-22.814911,-43.379011,Apartamento,Entire home/apt,4,1.0,1 bath,1.0,1.0,...,4.93,4.90,4.98,4.99,4.96,4.83,t,0.74,1.0,175.00
3,Barra da Tijuca,-23.010000,-43.344820,Apartamento,Entire home/apt,4,1.0,1 bath,1.0,1.0,...,5.00,4.82,5.00,5.00,5.00,5.00,f,0.60,1.0,125.00
4,Copacabana,-22.970696,-43.186048,Apartamento,Entire home/apt,6,2.0,2 baths,2.0,6.0,...,4.93,4.88,4.94,4.99,4.94,4.85,t,2.81,2.0,76.83
5,Leme,-22.963629,-43.172243,Apartamento,Entire home/apt,2,1.0,1 bath,1.0,2.0,...,4.83,4.87,5.00,5.00,5.00,4.70,f,0.73,1.0,140.50
6,Copacabana,-22.976303,-43.194838,Quarto Privado,Private room,2,1.0,1 shared bath,1.0,0.0,...,5.00,5.00,5.00,5.00,4.88,5.00,f,0.26,1.0,475.00
7,Botafogo,-22.940620,-43.181620,Quarto Privado,Private room,4,1.0,1 private bath,1.0,2.0,...,4.93,4.90,4.98,4.99,4.96,4.83,f,0.74,1.0,112.50
8,Copacabana,-22.979539,-43.191163,Apartamento,Entire home/apt,6,2.0,2 baths,3.0,4.0,...,4.67,4.83,4.67,4.83,4.67,4.33,f,0.19,2.0,166.67
9,Copacabana,-22.974970,-43.194830,Apartamento,Entire home/apt,3,3.0,3 baths,2.0,2.0,...,5.00,5.00,5.00,5.00,5.00,5.00,f,0.11,3.0,326.67


# 4. Exportando arquivo tratado


In [19]:
# No seu primeiro notebook
df.to_parquet('airbnb_rio_tratado.parquet', index=False)